Extended admin modify tutorials

Arcadius19 opened the pull request:
Enhanced the administrator dashboard where the user can add a new tutorial and modify or delete any existing one.
![screenshot 2](https://user-images.githubusercontent.com/22327474/43550307-0c0de488-95db-11e8-930c-726d67d20368.png)
![screenshot 3](https://user-images.githubusercontent.com/22327474/43550308-0c33d742-95db-11e8-98d7-95438f546e2a.png)
![screenshot 4](https://user-images.githubusercontent.com/22327474/43550309-0c53ccbe-95db-11e8-9981-3d0317ba0149.png)
![screenshot 5](https://user-images.githubusercontent.com/22327474/43550310-0c72e540-95db-11e8-9374-0122f1efbba7.png)

Arcadius19 committed:
Created initial routes for extended Administrator interface.
Added a side menu listing tutorials and an option to add a tutorial.

Arcadius19 committed:
Added a component for adding a new tutorial

Arcadius19 committed:
Changed a drawer (side menu) in admin interface to be of a type 'persistent'

Arcadius19 committed:
Implemented a functional ModifyTutorial component.
Added updateTutorial function to TryLinks Service.
Created a model of LinksTutorial with JSON serializer.

Arcadius19 committed:
Extracted styles for AddTutorial and ModifyTutorial components to a single file

Arcadius19 committed:
Adjusted a height of textareas for AddTutorial and ModifyTutorial

Arcadius19 committed:
Changed an order of lines to prevent from glitches when the tutorial's content is loaded.

Arcadius19 committed:
Added a functionality to delete a tutorial (functions in ModifyTutorial and TryLinks Service, and confirmation modals)

Arcadius19 committed:
Created UpdateListService to inform the Admin Component about a new tutorial or a tutorial modification so it can update the list on the side menu.

NickWu007 commented on lib/admin/admin.dart:
```diff
@@ -1,91 +1,65 @@
-      this.descEditor.getDoc().setValue('');
-      this.sourceEditor.getDoc().setValue('');
-    }
+  void updateList() {
```
What's this function for?

Arcadius19 replied:
Right, I used it when experimenting with the shared service between the Admin Components and its children, and then forgot to delete it.

NickWu007 commented on lib/admin/admin.html:
```diff
@@ -1,80 +1,28 @@
+            <material-icon icon="menu"></material-icon>
         </material-button>
+        <img class="tl-logo" src="img/logo.svg" alt="Links Logo">
+        <div class="tl-header-name">Try Links Admin Mode</div>
```
`Try Links`: Extra space

NickWu007 commented on lib/app_component.dart:
```diff
@@ -39,6 +39,6 @@ import 'package:client/admin/admin.dart';
       name: 'Tutorial',
       component: TutorialPageComponent),
-  const Route(path: '/admin', name: 'Admin', component: AdminPageComponent),
+  const Route(path: '/admin/...', name: 'Admin', component: AdminPageComponent),
```
What's this for?

Arcadius19 replied:
It indicated a non-terminal route and it's required for AdminPage to have child routes (/admin/add-tutorial and /modify-tutorial/:id)

NickWu007 commented on lib/admin/update-list-service.dart:
```diff
@@ -0,0 +1,15 @@
+
+  UpdateListService();
+
+  Stream get updateList => _updateListController.stream;
```
I would rename this as `updateHeadersStream`.

NickWu007 commented on lib/admin/update-list-service.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:angular/angular.dart';
+
+@Injectable()
+class UpdateListService {
```
Instead of naming this as a generic stream, I would use `UpdateHeadersService`.

NickWu007 commented on lib/admin/modify-tutorial/modify-tutorial.html:
```diff
@@ -0,0 +1,107 @@
+<h3 class="tutorial-panel-header">Modify a tutorial (id: {{tutorial.tutorialId}})</h3>
```
Don't think you need to expose the id here.

NickWu007 commented:
Good work Arek, just some small nits.

Btw make sure you update your local branch with the base MASTER branch. It seems like the background isn't as dim as it should be in the screenshots.

Arcadius19 committed:
Merge branch 'master' into ExtendedAdmin-ModifyTutorials

Arcadius19 committed:
Renamed UpdateListService to UpdateHeadersService and the stream within it.
Removed a tutorial ID form a modify-tutorial page.
Removed a space in TryLinks on Admin page.

Arcadius19 committed:
Changed a retrieval of a tutorial from POST to GET request.
Code populating textareas was moved into else-statement to prevent from an error displayed in a console if the tutorial is null

Arcadius19 committed:
Subscription for the updateHeaders event is cancelled when the component is destroyed so it does not throw an error about listening to the same stream again when other component instance is loaded
